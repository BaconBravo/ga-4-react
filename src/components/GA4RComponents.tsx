import React, { useState, useEffect } from 'react';
import GA4React from '..';
adimport { GA4ReactResolveInterface } from '../lib/gtagModels';

export interface IGAReactConfig {
  send_page_view: boolean;
  groups: string;
}

export interface IGA4R {
  code: string;
  config?: IGAReactConfig;
  additionalCode?: Array<string>;
  children?: any;
}

export const GA4R: React.FC<IGA4R> = ({
  code,
  config,
  additionalCode,
  children,
}) => {
  const [components, setComponents] = useState<any>(children);

  useEffect(() => {
    const ga4manager = new GA4React(`${code}`, config, additionalCode);
    ga4manager.initialize().then(
      (ga4: GA4ReactResolveInterface) => {
        setComponents(
          React.Children.map(children, (child: React.ReactChildren, index) => {
            if (!React.isValidElement(child)) {
              return <React.Fragment>{child}</React.Fragment>;
            } else {
              //@ts-ignore
              if (child.type && typeof child.type.name !== 'undefined') {
                return React.cloneElement(child, {
                  //@ts-ignore
                  ga4: ga4,
                  index,
                });
              } else {
                return child;
              }
            }
          })
        );
      },
      err => {
        console.error(err);
      }
    );
  }, []);

  return <>{components}</>;
};

export default GA4R;
