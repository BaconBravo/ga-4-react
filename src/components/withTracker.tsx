import React, { useEffect } from 'react';
import GA4React from '..';
import { GA4ReactResolveInterface } from '../lib/gtagModels';

export interface GA4WithTrackerComponentInterface {
  path: string | Location;
  location?: string | Location;
  title?: string;
  gaCode?: string;
}

export function withTracker(
  MyComponent: React.FC<any>
): React.FC<GA4WithTrackerComponentInterface> {
  return (props: GA4WithTrackerComponentInterface & any) => {
    const { path, location, title, gaCode } = props;
    useEffect(() => {
      switch (GA4React.isInitialized()) {
        case true:
          const ga4 = GA4React.getGA4React();
          if (ga4) {
            ga4.pageview(path, location, title);
          }
          break;
        default:
        case false:
          const ga4react = new GA4React(gaCode);
          ga4react.initialize().then(
            (ga4: GA4ReactResolveInterface) => {
              ga4.pageview(path, location, title);
            },
            (err: Error) => {
              console.error(err);
            }
          );
          break;
      }
    });
    return <MyComponent {...props} />;
  };
}

export default withTracker;
