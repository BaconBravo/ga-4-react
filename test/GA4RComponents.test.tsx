import React from 'react';
import { render } from '@testing-library/react';
import { GA4R } from './../src/components/GA4RComponents';

describe('GA4R Components', () => {
  it('render', async done => {
    const { container } = render(
      <GA4R code="GA-CODE">
        <div>Try</div>
      </GA4R>
    );

    console.log(container.innerHTML);

    setTimeout(() => {
      expect(container.innerHTML).toMatchSnapshot();
      done();
    }, 100);
  });
});
