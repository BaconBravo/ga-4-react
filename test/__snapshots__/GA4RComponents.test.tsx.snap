// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GA4R Components render 1`] = `""`;
